File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
  def lengthOfLongestSubstring(self, s: str) -> int:
    if len(s) == 0: return 0
    charDict = {}
    lastRepeat = 0
    result = 0
    for i in range(len(s)):
      if s[i] in charDict and charDict[s[i]] >= lastRepeat:
        lastRepeat = charDict[s[i]] + 1
      else:
        result = max(result, i + 1 - lastRepeat)
      charDict[s[i]] = i
    return result

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cs
public class Solution {
  public int LengthOfLongestSubstring(string s) {
    var cache = new Dictionary<char, int>();
    int lastRepeat = 0;
    int result = 0;
    for (int i = 0; i < s.Length; i++) {
      char ch = s[i];
      if (cache.ContainsKey(ch) && cache[ch] >= lastRepeat) {
        lastRepeat = cache[ch] + 1;
      } else {
        result = Math.Max(result, i + 1 - lastRepeat);
      }
      cache[ch] = i;
    }
    
    return result;
  }
}